### PR TITLE
GH-39049: [C++] Use Cast() instead of CastTo() for Dictionary Scalar in test

### LINKED ISSUE #39049: [C++] Add Compute Kernel for Casting from String to Dictionary
### Describe the enhancement requested

This is a sub-issue of the issue mentioned below.

* #35560 

As we can see in the parent issue, we can find tables organizing the related functions.

| file name         | code snippet                                                                               |
|-------------------|--------------------------------------------------------------------------------------------|
| partition_test.cc | `ASSERT_OK_AND_ASSIGN(auto dict_hello, MakeScalar("hello")->CastTo(DictStr("")->type()));` |
| partition_test.cc | `ASSERT_OK_AND_ASSIGN(auto dict_hello, MakeScalar("hello")->CastTo(dict_type));`           |
| scalar_test.cc    | `ASSERT_OK_AND_ASSIGN(auto cast_alpha, alpha->CastTo(ty));`                                |

> The functions result in "Unsupported cast from string to dictionary using function cast_dictionary" error.

We need to implement casting from `String` to `Dictionary`. 

#### partition_test.cc

Snippet to reproduce

```diff
TEST_F(TestPartitioning, DirectoryPartitioningFormatDictionary) {

  ...

- ASSERT_OK_AND_ASSIGN(auto dict_hello, MakeScalar("hello")->CastTo(DictStr("")->type()));
+ ASSERT_OK_AND_ASSIGN(auto dict_hello, Cast(MakeScalar("hello"), DictStr("")->type()));
}

TEST_F(TestPartitioning, DirectoryPartitioningFormatDictionaryCustomIndex) {

  ...

- ASSERT_OK_AND_ASSIGN(auto dict_hello, MakeScalar("hello")->CastTo(dict_type));
+ ASSERT_OK_AND_ASSIGN(auto dict_hello, Cast(MakeScalar("hello"), dict_type));
```

#### scalar_test.cc

Snippet to reproduce

```diff
TEST(TestDictionaryScalar, Cast) {

  ...

      // Cast string to dict(..., string)
-     ASSERT_OK_AND_ASSIGN(auto cast_alpha, alpha->CastTo(ty));
-     ASSERT_OK(cast_alpha->ValidateFull());
+     ASSERT_OK_AND_ASSIGN(auto cast_alpha, Cast(alpha, ty));
+     const auto& scalar = cast_alpha.scalar();
+     ASSERT_OK(scalar->ValidateFull());
      ASSERT_OK_AND_ASSIGN(
          auto roundtripped_alpha,
-         checked_cast<const DictionaryScalar&>(*cast_alpha).GetEncodedValue());
+         checked_cast<const DictionaryScalar&>(*scalar).GetEncodedValue());

      ...

      // dictionaries differ, though encoded values are identical
-     ASSERT_FALSE(alpha_dict.Equals(*cast_alpha));
+     ASSERT_FALSE(alpha_dict.Equals(*scalar));
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by llama90**
### Rationale for this change

Remove legacy code

### What changes are included in this PR?

Replace the legacy scalar CastTo implementation for Dictionary Scalar in test.

### Are these changes tested?

Yes. It is passed by existing test cases.

### Are there any user-facing changes?

No.

* Closes: #39049

### CODE REVIEW

**Comment by llama90** [comment]
@kou Would it be possible for you to review this PR? I anticipate that once this PR and the GLib PR are merged, we will be able to proceed with the task of deprecating `CastTo`. #39182 

Thank you as always.

File: cpp/src/arrow/scalar_test.cc [hunk 1]
```diff
@@ -1479,11 +1479,12 @@ TEST(TestDictionaryScalar, Cast) {
       auto alpha =
           dict->IsValid(i) ? MakeScalar(dict->GetString(i)) : MakeNullScalar(utf8());
       // Cast string to dict(..., string)
-      ASSERT_OK_AND_ASSIGN(auto cast_alpha, alpha->CastTo(ty));
-      ASSERT_OK(cast_alpha->ValidateFull());
+      ASSERT_OK_AND_ASSIGN(auto cast_alpha, Cast(alpha, ty));
+      const auto& scalar = cast_alpha.scalar();
```
Comment (kou) [inline cpp/src/arrow/scalar_test.cc:1483]: How about keeping `cast_alpha` name? It's more meaningful than `scalar`.

```suggestion
      ASSERT_OK_AND_ASSIGN(auto cast_alpha_datum, Cast(alpha, ty));
      const auto& cast_alpha = cast_alpha_datum.scalar();
```

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc [hunk 2]
```diff
@@ -77,17 +85,23 @@ Status CastToDictionary(KernelContext* ctx, const ExecSpan& batch, ExecResult* o
...
-  AddCommonCasts(Type::DICTIONARY, kOutputTargetType, func.get());
-  ScalarKernel kernel({InputType(Type::DICTIONARY)}, kOutputTargetType, CastToDictionary);
+template <typename SrcType>
+void AddDictionaryCast(CastFunction* func) {
+  ScalarKernel kernel;
+  kernel.exec = CastToDictionary;
+  kernel.signature =
+      KernelSignature::Make({InputType(SrcType::type_id)}, kOutputTargetType);
   kernel.null_handling = NullHandling::COMPUTED_NO_PREALLOCATE;
-  kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:86]: Can we keep this for `DictionaryType`?

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc [hunk 3]
```diff
@@ -44,6 +44,14 @@ Status CastToDictionary(KernelContext* ctx, const ExecSpan& batch, ExecResult* o
   }
 
   std::shared_ptr<ArrayData> in_array = batch[0].array.ToArrayData();
+
+  // If the input type is STRING, it is first encoded as a dictionary to facilitate
+  // processing. This approach allows the subsequent code to uniformly handle STRING
+  // inputs as if they were originally provided in dictionary format. Encoding as a
+  // dictionary helps in reusing the same logic for dictionary operations.
+  if (batch[0].type()->id() == Type::STRING) {
+    in_array = DictionaryEncode(batch[0].array.ToArrayData())->array();
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:53]: Can we use `in_array` here?

```suggestion
    in_array = DictionaryEncode(in_array)->array();
```

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc [hunk 4]
```diff
@@ -77,17 +85,23 @@ Status CastToDictionary(KernelContext* ctx, const ExecSpan& batch, ExecResult* o
...
-  kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
+  DCHECK_OK(func->AddKernel(SrcType::type_id, std::move(kernel)));
+}
 
-  DCHECK_OK(func->AddKernel(Type::DICTIONARY, std::move(kernel)));
+std::vector<std::shared_ptr<CastFunction>> GetDictionaryCasts() {
+  auto cast_dict = std::make_shared<CastFunction>("cast_dictionary", Type::DICTIONARY);
+  AddCommonCasts(Type::DICTIONARY, kOutputTargetType, cast_dict.get());
+  AddDictionaryCast<DictionaryType>(cast_dict.get());
+  AddDictionaryCast<StringType>(cast_dict.get());
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:102]: Can we use `arrow::StringTypes()` for all string types (`StringType` and `LargeStringType`)?

(We may want to add support for `BinaryType` and `LargeBinaryType` as a separated PR.)

File: cpp/src/arrow/scalar_test.cc (hunk 1 above)
Comment (llama90) [reply cpp/src/arrow/scalar_test.cc]: Fix it.

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc (hunk 2 above)
Comment (llama90) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:86]: too.

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc (hunk 3 above)
Comment (llama90) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc]: I agree.

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc (hunk 4 above)
Comment (llama90) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:103]: I understand. You're right. Can we address that part as a known issue in a new PR? I would like to first finish the work on removing the Legacy Cast.

Comment (kou) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:103]: It makes sense.

Comment (llama90) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:103]: I created the issue.

- #39463

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc [hunk 5]
```diff
@@ -77,17 +85,24 @@ Status CastToDictionary(KernelContext* ctx, const ExecSpan& batch, ExecResult* o
...
-  auto func = std::make_shared<CastFunction>("cast_dictionary", Type::DICTIONARY);
-
-  AddCommonCasts(Type::DICTIONARY, kOutputTargetType, func.get());
-  ScalarKernel kernel({InputType(Type::DICTIONARY)}, kOutputTargetType, CastToDictionary);
+template <typename SrcType>
+void AddDictionaryCast(CastFunction* func) {
+  ScalarKernel kernel;
+  kernel.exec = CastToDictionary;
+  kernel.signature =
+      KernelSignature::Make({InputType(SrcType::type_id)}, kOutputTargetType);
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:93]: How about using the `ScalarKernel()` constructor like the original code?

```cpp
  ScalarKernel kernel({InputType(SrcType::type_id)}, kOutputTargetType, CastToDictionary);
```

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc (hunk 5 above)
Comment (llama90) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc]: I have made changes and identified the following code snippets. Is it recommended to use the `ScalarKernel()` constructor?

https://github.com/apache/arrow/blob/1c1fa746f8d1266f9210d9c0e1bbab074c5dd4e0/cpp/src/arrow/compute/kernels/scalar_cast_nested.cc#L148-L151

https://github.com/apache/arrow/blob/1c1fa746f8d1266f9210d9c0e1bbab074c5dd4e0/cpp/src/arrow/compute/kernels/scalar_cast_temporal.cc#L495-L497

...

Would it be advisable to open a new issue for these changes if they are recommended?

Comment (kou) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc]: Oh. I didn't notice it.
The constructor is a shortcut of these existing codes: https://github.com/apache/arrow/blob/1c1fa746f8d1266f9210d9c0e1bbab074c5dd4e0/cpp/src/arrow/compute/kernel.h#L496-L498
So I think the constructor is recommended.

> Would it be advisable to open a new issue for these changes if they are recommended?

Yes, please. We can discuss there what API is recommended.

**Pull request merged**